WikiWatershed/model-my-watershed
# Custom Weather: Usage Across Scenarios

## Overview

Adds support for duplicating scenarios with custom weather applied. The custom weather modification is copied over, and the same file is referenced.

During deletion, we check to see if the file is used by any other scenario. If so, we ask the user to either delete those scenarios or set them to use Available data and try again.

Also updates the UI to always save on valid states, and remove the Cancel / Save button in favor of a Done button.

See commit messages for details.

Connects #3309

### Notes

This includes a fix for an error specified in #3312. I'm leaving that card open to gather feedback from clients, but wanted to get this in before giving it over to them for testing.

## Testing Instructions

* Check out this branch and `bundle --debug`
* Go to [:8000/](http://localhost:8000) and create a new MapShed project
* Add changes to it. Rename "New Scenario" to "Original".
    ![image](https://user-images.githubusercontent.com/1430060/82268355-cc20a380-993c-11ea-861f-3081943da65d.png)
* Open the Weather Data dialog
    ![image](https://user-images.githubusercontent.com/1430060/82268424-f83c2480-993c-11ea-94eb-fbd88af34280.png)
  - [x] Ensure you see a Done button instead of Cancel / Save
* Add a valid custom weather file.
    ![image](https://user-images.githubusercontent.com/1430060/82268508-2de10d80-993d-11ea-813d-cc760b943525.png)
  - [x] Ensure the scenario is updated immediately on successful save
* Click Done to dismiss the dialog. Duplicate the scenario.
    ![image](https://user-images.githubusercontent.com/1430060/82268553-4ea96300-993d-11ea-848c-03d847cdda30.png)
  - [x] Ensure there is a new scenario called "Copy of Original"
* Open its Weather Data
    ![image](https://user-images.githubusercontent.com/1430060/82268597-684aaa80-993d-11ea-9b39-70df76c42f80.png)
  - [x] Ensure it is the same as Original
* Try to delete the weather file
    ![image](https://user-images.githubusercontent.com/1430060/82268736-bd86bc00-993d-11ea-9490-7224a7db2b6f.png)
  - [ ] Ensure you are stopped from doing so
* Go back to Original and switch it to Available Data
    ![image](https://user-images.githubusercontent.com/1430060/82268809-dee7a800-993d-11ea-8c0e-ee496dd00e65.png)
* Go back to Copy of Original and delete the file
    ![image](https://user-images.githubusercontent.com/1430060/82268839-f1fa7800-993d-11ea-8911-47d8f686bda0.png)
  - [x] Ensure there are no errors this time
* Go back to Original and open Weather Data
    ![image](https://user-images.githubusercontent.com/1430060/82268866-08083880-993e-11ea-8990-55291b554e85.png)
  - [ ] Ensure no custom data file is shown
